### ISSUE TITLE
Split read/write types for `@field`s

### ISSUE THREAD

**Issue by justarandomgeek** [EmmyLuaLs/emmylua-analyzer-rust#1129]
In Factorio's API (and presumably many others?) we have a large number of API objects with properties that accept assignment from a wide union of types but only ever return one specific member of that union when read. In particular, there are many that accept writing nil to reset them to default states, but will never read as nil.

Unfortunately, since we can only give the `@field` a single type, we currently have to give it the wider of the two, which results in many spurious warnings about nil checks and type tests on values read from these affected properties.

Would it be possible to have separate types for `@field`s to validate against for read/write access?

**Comment by justarandomgeek** [comment]
aside: this would also improve support for both read-only and write-only fields (which we have a few of each of) by simply specifying the other side as `never`!

**Comment by xuhuanzy** [comment]
Perhaps we need to add a new annotation tag for support, but not right now.

**Comment by PennyJim** [comment]
My idea for the tag annotation is to have it be described like this:
```lua
---@class example
---
---@field (read) readwrite_field_one A
---@field (write) readwrite_field_one A|B|C
---
---@field (split) readwrite_field_two A, A|B|C -- I wouldn't be heartbroken to lose this syntax
---
---@field (read) readonly_field A   -- Implicitly sets the unset `readonly_field` write to `never`
---@field (write) writeonly_field B -- Implicitly sets the unset `writeonly_field` read to `never`
```

I see that `@readonly` does exist for variables, but I see no way to get it to work for class fields where I think it makes a similar amount of sense, if not more, but it being a `@` tag feels uncomfortable within a class definition.

**Comment by justarandomgeek** [comment]
while looking around at other definitions, we also found the `@attribute field_accessor` which does not currently do this but seems like it might also be a good fit if it were hooked up for get/set types from the functions (return type of the getter, arg type of the setter)!
